Zend/zend_multiply.h

```diff
@@ -19,10 +19,32 @@
 
 /* $Id$ */
 
+#include "zend_portability.h"
+
 #ifndef ZEND_MULTIPLY_H
 #define ZEND_MULTIPLY_H
 
-#if (defined(__i386__) || defined(__x86_64__)) && defined(__GNUC__)
+#if __has_builtin(__builtin_smull_overflow) && SIZEOF_LONG == SIZEOF_ZEND_LONG
```

Comment:
Both GCC and clang support some of these intrinsics. I specifically chose the subset that I know both GCC and clang support (clang has a few extra ones).

As Xinchen points out, we define `__has_builtin` as a fallback in `zend_portability.h`, which is why I added an `#include` for it.